model_templates/training/python3_pytorch/custom.py

```diff
@@ -47,15 +48,42 @@ def fit(
     -------
     Nothing
     """
+
+    preprocessor, cols = build_preprocessor(X)
+    preprocessor.fit(X)
+    X_preprocessed = preprocessor.transform(X)
+    X_preprocessed = pd.DataFrame(X_preprocessed, columns=cols)
+
     # Feel free to delete which ever one of these you aren't using
     if class_order:
-        estimator, optimizer, criterion = build_classifier(X)
-        train_classifier(X, y, estimator, optimizer, criterion)
+        estimator, optimizer, criterion = build_classifier(X_preprocessed)
+        train_classifier(X_preprocessed, y, estimator, optimizer, criterion)
         artifact_name = "torch_bin.pth"
     else:
-        estimator, optimizer, criterion = build_regressor(X)
-        train_regressor(X, y, estimator, optimizer, criterion)
+        estimator, optimizer, criterion = build_regressor(X_preprocessed)
+        train_regressor(X_preprocessed, y, estimator, optimizer, criterion)
         artifact_name = "torch_reg.pth"
 
     # NOTE: We currently set a 10GB limit to the size of the serialized model
     save_torch_model(estimator, output_dir, artifact_name)
+
+
+def transform(data, model):
+    """
+    Apply the same preprocessing on prediction as on fit
+
+    Parameters
+    ----------
+    data: pd.DataFrame
+    model: object, the deserialized model
+
+    Returns
+    -------
+    pd.DataFrame
+    """
+    # Execute any steps you need to do before scoring
+    # Remove target columns if  they're in the dataset
+    preprocessor, cols = build_preprocessor(data)
+    preprocessor.fit(data)
```

Comment:
This doesn't seem right, the preprocessor shouldn't be rebuilt at transform time

Reply:
oh good call - so we may need to wrap both the preprocessor and modeler in a separate (sklearn-style?) estimator object that we can pickle/unpickle as one unit...or, I guess, save two artifacts in fit and write another custom hook to unpack them?

Reply:
An sklearn pipeline that contains all preprocessing and the estimator would be ideal

Reply:
yeah- I'm not sure that there's a nice plug-and-play way to do that with pytorch

Reply:
This has a good example: https://www.kaggle.com/graymant/pytorch-regression-with-sklearn-pipelines
Have to extend the sklearn BaseEstimator class

Reply:
....I'm going to punt on this task (making a new ticket to figure out a good framework for this) and just put in a subset transformer to keep only numerical features for now. WIP of the sklearn estimator is [here](https://github.com/datarobot/datarobot-user-models/tree/jlpolit/pytorch-sklearn-estimator-wip) but it still needs some work